### ISSUE TITLE
Seed for dropout in LSTM - Difference in model(X) and model.predict(X)

### ISSUE THREAD

**Issue by lakshaykc** [tensorflow/tensorflow#41610]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.0.0-rc2-26-g64c3d382ca 2.0.0
- Python version: 3.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
Outputs for LSTM layer in tensorflow when using model(X) and model.predict(X) differ when using dropout.

Let's call the output of model(X) as Fwd Pass and model.predict(X) as Prediction

For a regular dropout layer, we can specify the seed but LSTM layer doesn't have such an argument. I'm guessing this is causing the difference between these Fwd Pass and Prediction.

In the following code sample, if dropout=0.4, these the outputs are different but when dropout=0.0 they match exactly. This makes me believe that every evaluation is using a different operation level seed but there is no way to set that.

PS: I want to use dropout during inference, so that is by design.

**Describe the expected behavior**
**Fwd Pass** and **Prediction** should be exactly the same when using dropout.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

```
import os
import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.initializers import GlorotUniform

SEED = 200
HIDDEN_UNITS = 4
N_OUTPUTS = 1
N_INPUTS = 4
BATCH_SIZE = 4
N_SAMPLES = 4

np.random.seed(SEED)
tf.random.set_seed(SEED)


# Simple LSTM Model
def my_model():
    inputs = x = keras.Input(shape=(N_INPUTS, 1))
    initializer = GlorotUniform(seed=SEED)

    x = layers.LSTM(HIDDEN_UNITS,
                     kernel_initializer=initializer,
                     recurrent_dropout=0.0,
                     dropout=0.4,
                     # return_sequences=True,
                     use_bias=False)(x, training=True)

    output = x
    model = keras.Model(inputs=inputs, outputs=[output])
    return model

# Create Sample Data
# Target Function
def f_x(x):
    y = x[:, 0] + x[:, 1] ** 2 + np.sin(x[:, 2]) + np.sin(x[:, 3] ** 3)
    y = y[:, np.newaxis]
    return y

# Generate random inputs
d = np.linspace(0.1, 1, N_SAMPLES)
X = np.transpose(np.vstack([d*0.25, d*0.5, d*0.75, d]))
X = X[:, :, np.newaxis]
Y = f_x(X)

# PRINT FWD PASS
model = my_model()
n_out = model(X).numpy()
print('FWD PASS:')
print(n_out, '\n')

# PRINT PREDICT OUTPUT
print('PREDICT:')
out = model.predict(X)
print(out)
```

**Output (dropout=0.4) - do not match**

```
FWD PASS:
[[ 0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.        ]
 [ 0.0526864  -0.13284351  0.02326298 -0.30357683]
 [ 0.06297918 -0.14084947  0.02214929 -0.44425806]] 

PREDICT:
[[ 0.00975818 -0.029404    0.00678372 -0.03232396]
 [ 0.0347842  -0.0974849   0.01938616 -0.15696262]
 [ 0.          0.          0.          0.        ]
 [ 0.06297918 -0.14084947  0.02214929 -0.44425806]]
```

**Output (dropout=0.0) - no dropout, outputs match**

```
FWD PASS:
[[ 0.00593475 -0.01799661  0.00424165 -0.01876264]
 [ 0.02226446 -0.06519517  0.01399653 -0.08595844]
 [ 0.03620889 -0.10084937  0.01987283 -0.1663805 ]
 [ 0.0475584  -0.12453148  0.02269932 -0.2541136 ]] 

PREDICT:
[[ 0.00593475 -0.01799661  0.00424165 -0.01876264]
 [ 0.02226446 -0.06519517  0.01399653 -0.08595844]
 [ 0.03620889 -0.10084937  0.01987283 -0.1663805 ]
 [ 0.0475584  -0.12453148  0.02269932 -0.2541136 ]]
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by lakshaykc** [comment]
for further clarification, I am calling the LSTM layer with the argument `training=True`, so that dropout is enabled when `model.predict()` is called

**Comment by Saduf2019** [comment]
i am able to replicate this issue reported, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/d4afdd07e202c47b5b4533e66a6b61eb/untitled292.ipynb)

**Comment by sushreebarsa** [comment]
Was able to replicate the issue in TF 2.6.0-dev20210530,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/25b97830f9b53c0a386d96eb9a1305f6/untitled114.ipynb#scrollTo=VzPWy9TmOmC0)..Thanks !

**Comment by kumariko** [comment]
@lakshaykc I tried to run your code on TF 2.7 and got different result, please find the gist [here](https://colab.research.google.com/gist/kumariko/4be9813a59e5a1845f5572cd192b3d25/untitled114.ipynb#scrollTo=pWgavkVhOn85) and confirm the same.

**Comment by shashankdeshpande** [comment]
Issue is not resolved yet, getting different results if we specify dropout even global seed has been already set with below commands - 
```python
import os
import numpy as np
import tensorflow as tf
import random as python_random

global_seed = 42
np.random.seed(global_seed)
python_random.seed(global_seed)
tf.random.set_seed(global_seed)
tf.keras.utils.set_random_seed(global_seed)
os.environ['PYTHONHASHSEED'] = str(global_seed)
```
LSTM
```python
tf.keras.layers.LSTM(
  64,
  activation='elu',
  dropout=0.2,
  return_state=False,
  return_sequences=True
)
```
dropout parameter in tf.keras.layers.LSTM requires float value, global seed is not getting applied here.
Am I missing something?
